File: WPF/Controls/PopupPanel.vb
Namespace Common.Controls

    Public Class PopupPanel
        Inherits Panel

        Protected Overrides Function MeasureOverride(ByVal AvailableSize As Size) As Size
            Dim MaxWidth = 0.0
            Dim MaxHeight = 0.0

            For Each C As UIElement In Me.Children
                C.Measure(AvailableSize)
                Dim Sz As Size = C.DesiredSize

                If MaxHeight < Sz.Height Then
                    MaxHeight = Sz.Height
                End If
                If MaxWidth < Sz.Width Then
                    MaxWidth = Sz.Width
                End If
            Next

            Return New Size(MaxWidth, MaxHeight)
        End Function

        Protected Overrides Function ArrangeOverride(FinalSize As Size) As Size
            For Each C As UIElement In Me.Children
                Dim Popup = TryCast(C, Popup)
                Dim Rect As Rect?

                If Popup IsNot Nothing Then
                    Rect = Popup.ArrangeCallBack?.Invoke(Me, FinalSize, Popup.DesiredSize)
                End If

                If Not Rect.HasValue Then
                    Rect = New Rect(FinalSize)
                End If

                C.Arrange(Rect.Value)
            Next

            Return FinalSize
        End Function

    End Class

End Namespace

File: WPF/Controls/PopupPanel.cs
using System.Windows;
using System.Windows.Controls;

namespace Ks.Common.Controls
{
        public class PopupPanel : Panel
        {
            protected override Size MeasureOverride(Size AvailableSize)
            {
                var MaxWidth = 0.0;
                var MaxHeight = 0.0;

                foreach (UIElement C in this.Children)
                {
                    C.Measure(AvailableSize);
                    var Sz = C.DesiredSize;

                    if (MaxHeight < Sz.Height)
                        MaxHeight = Sz.Height;
                    if (MaxWidth < Sz.Width)
                        MaxWidth = Sz.Width;
                }

                return new Size(MaxWidth, MaxHeight);
            }

            protected override Size ArrangeOverride(Size FinalSize)
            {
                foreach (UIElement C in this.Children)
                {
                    var Popup = C as Popup;
                    var Rect = default(Rect?);

                    if (Popup != null)
                        Rect = Popup.ArrangeCallBack?.Invoke(this, FinalSize, Popup.DesiredSize);

                    if (!Rect.HasValue)
                        Rect = new Rect(FinalSize);

                    C.Arrange(Rect.Value);
                }

                return FinalSize;
            }
        }
    }
